# Домашнее задание №3.

*Выполнила: Книс Надежда*

## Задание 1

In [1]:
import pandas as pd
df = pd.read_stata('DZ_var10.dta') 

In [2]:
df

idind  year                                         region       h5  \
0          1  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
1      11293  2012       Ленинградская область, Волосовский район  МУЖСКОЙ   
2      24101  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
3          5  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
4          7  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
...      ...   ...                                            ...      ...   
22529  47958  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   
22530  47959  2012  Новосибирская область, Бердский район (с 2003  МУЖСКОЙ   
22531  47960  2012  Новосибирская область, Бердский район (с 2003  МУЖСКОЙ   
22532  47961  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   
22533  47962  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   

      age monthly_wage  married  higher_educ  
0      39      43000.0      0.0          1.0  
1      17          NaN      0.0          0.0  
2       9          NaN      NaN          0.0  
3      28      45000.0      1.0          0.0  
4      68          NaN      0.0          0.0  
...    ..          ...      ...          ...  
22529  16          NaN      0.0          0.0  
22530   7          NaN      NaN          0.0  
22531  44      20000.0      1.0          0.0  
22532  41      10000.0      1.0          0.0  
22533  21      10000.0      0.0          0.0  

[22534 rows x 8 columns]

## Задание 2

In [8]:
df.rename(columns= {'h5':'gender'}, inplace=True)

In [9]:
df

idind  year                                         region   gender  \
0          1  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
1      11293  2012       Ленинградская область, Волосовский район  МУЖСКОЙ   
2      24101  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
3          5  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
4          7  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
...      ...   ...                                            ...      ...   
22529  47958  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   
22530  47959  2012  Новосибирская область, Бердский район (с 2003  МУЖСКОЙ   
22531  47960  2012  Новосибирская область, Бердский район (с 2003  МУЖСКОЙ   
22532  47961  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   
22533  47962  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   

      age monthly_wage  married  higher_educ  
0      39      43000.0      0.0          1.0  
1      17          NaN      0.0          0.0  
2       9          NaN      NaN          0.0  
3      28      45000.0      1.0          0.0  
4      68          NaN      0.0          0.0  
...    ..          ...      ...          ...  
22529  16          NaN      0.0          0.0  
22530   7          NaN      NaN          0.0  
22531  44      20000.0      1.0          0.0  
22532  41      10000.0      1.0          0.0  
22533  21      10000.0      0.0          0.0  

[22534 rows x 8 columns]

## Задание 3.1

**Какую среднюю заработную плату получают мужчины и женщины в выборке? Запишите команду. Выпишите значение в тысячах рублей.**

In [11]:
df.dropna(inplace=True)

In [12]:
df

idind  year                                         region   gender  \
0          1  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
3          5  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
7      21443  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
8          3  2012       Ленинградская область, Волосовский район  ЖЕНСКИЙ   
9      25025  2012       Ленинградская область, Волосовский район  МУЖСКОЙ   
...      ...   ...                                            ...      ...   
22525  47954  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   
22527  47956  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   
22531  47960  2012  Новосибирская область, Бердский район (с 2003  МУЖСКОЙ   
22532  47961  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   
22533  47962  2012  Новосибирская область, Бердский район (с 2003  ЖЕНСКИЙ   

      age monthly_wage  married  higher_educ  
0      39      43000.0      0.0          1.0  
3      28      45000.0      1.0          0.0  
7      52      15000.0      0.0          0.0  
8      57      12000.0      0.0          0.0  
9      35      28000.0      1.0          0.0  
...    ..          ...      ...          ...  
22525  39      30000.0      1.0          0.0  
22527  37      10000.0      1.0          1.0  
22531  44      20000.0      1.0          0.0  
22532  41      10000.0      1.0          0.0  
22533  21      10000.0      0.0          0.0  

[9491 rows x 8 columns]

In [16]:
df.dtypes

idind              int32
year               int16
region          category
gender          category
age             category
monthly_wage    category
married          float64
higher_educ      float32
dtype: object

In [20]:
df['monthly_wage'] = df['monthly_wage'].astype('float')

In [48]:
df['age'] = df['age'].astype('int')

In [21]:
a = df['monthly_wage'].mean()
print(a/1000)

19.10615064060689


## Задание 3.2

**Каков минимум и максимум заработной платы в выборке? Выпишите значение в тысячах рублей.**

In [24]:
min = df['monthly_wage'].min()/1000
max = df['monthly_wage'].max()/1000
print('min:', min, 'max:', max)

min: 1.0 max: 400.0


## Задание 4.1

**Создайте новую переменную для логарифма заработной платы. Выпишите среднее значение для выборки. Округлите до десятых.**

In [25]:
import numpy as np

In [27]:
df['ln_wage'] = np.log(df['monthly_wage'])

In [29]:
round(df['ln_wage'].mean(),1)

9.6

## Задание 4.2

**Создайте новую переменную для квадрата возраста. Выпишите среднее значение для выборки. Округлите до десятых.**

In [49]:
df['sq_age'] = np.square(df['age'])

In [50]:
round(df['sq_age'].mean(),1)

1748.0

## Задание 5

**Какая доля выборки имеет высшее образование? Запишите процент.**

In [47]:
a = df[df['higher_educ']==1.0].shape[0]
b = df.shape[0]
print(a/b*100)

30.79759772415973


## Задание 6

**Оцените МНК-модель, где в качестве зависимой переменной выступает логарифм заработной платы, а в качестве объясняющих переменных – пол, возраст, квадрат возраста, высшее образование, семейное положение. Запишите команду.**

In [51]:
import statsmodels.api as sm

In [52]:
df2  = pd.get_dummies(df, columns=['gender', 'married', 'higher_educ'], drop_first=True)
df2['gender_ЖЕНСКИЙ'] = df2['gender_ЖЕНСКИЙ'].astype('int')
df2['higher_educ_1.0'] = df2['higher_educ_1.0'].astype('int')
df2['married_1.0'] = df2['married_1.0'].astype('int')
X = df2[['gender_ЖЕНСКИЙ', 'age', 'sq_age', 'higher_educ_1.0', 'married_1.0']]
Y = df2['ln_wage']
X  =sm.add_constant(X)
res = sm.OLS(Y,X).fit()
res.summary()


C:\Users\Honor\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                ln_wage   R-squared:                       0.177
Model:                            OLS   Adj. R-squared:                  0.176
Method:                 Least Squares   F-statistic:                     407.0
Date:                Sat, 30 Sep 2023   Prob (F-statistic):               0.00
Time:                        02:11:45   Log-Likelihood:                -9222.3
No. Observations:                9491   AIC:                         1.846e+04
Df Residuals:                    9485   BIC:                         1.850e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const               8.9937      0.068    132.383      0.000       8.861       9.127
gender_ЖЕНСКИЙ     -0.4489      0.014    -33.072      0.000      -0.476      -0.422
age                 0.0441      0.003     12.770      0.000       0.037       0.051
sq_age             -0.0006   4.04e-05    -14.494      0.000      -0.001      -0.001
higher_educ_1.0     0.3930      0.014     27.415      0.000       0.365       0.421
married_1.0        -0.0107      0.016     -0.682      0.495      -0.041       0.020
==============================================================================
Omnibus:                      158.750   Durbin-Watson:                   1.404
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              276.809
Skew:                          -0.126   Prob(JB):                     7.79e-61
Kurtosis:                       3.798   Cond. No.                     2.12e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.12e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## Задание 7

**Является ли получившаяся модель адекватной? Выпишите значение теста на адекватность и p-value.**

Значение теста на адекватность: F-statistic =  407.0   

Один из вариантов проверки модели на адекватность - это значимость F-статистики. Нулевая гипотеза теста об адекватности - это  равенство всех коэффициентов модели нулю. Соответственно, если она будет отвергаться, то модель является адекватной. Поэтому в нашем случае смотрим на p-value = Prob (F-statistic) = 0.00, он меньше, чем 0.01 или 0.05 (часто используемые уровни значимости), а значит, наша модель адекватна, поскольку нулевая гипотеза отвергается.

## Задание 8

**Какую долю дисперсии заработных плат объясняет ваша регрессионная модель? Выпишите значение.**

Долю объясняемой дисперсии можно посмотреть по коэффициенту детерминации R^2.  

R-squared = 0.177.  

Adj. R-squared = 0.176 (скорректированный коэф-т детерминации на случай увеличения объясняющих переменных в модели, т.е. чтобы в этом случае от роста числа переменных не росла объясняемость)

## Задание 9

**Как высшее образование связано с уровнем заработных плат? Является ли коэффициент значимым? Проинтерпретируйте коэффициент увеличения/сокращения заработной платы.**

Коэфициент при переменной высшего образования = 0.3930, значит, при увеличении уровня высшего образования на 1 ед. логарифм уровня зарплат растет на 0.3930 единиц. Положительная (прямая) зависимость.  

P>|t|  = 0.00 => p-value = 0.00, что меньше любого уровня значимости.
Значит, коэффициент является значимым.

## Задание 10

**Проведите тест Бройша-Пагана на выявление гетероскедастичности и выпишите значение p-value. Проинтерпретируйте результат. Возникает ли в модели проблема гетероскедастичности?**

In [59]:
import statsmodels.stats.api as sms
from statsmodels.compat import lzip

In [60]:
names = ['Lagrange multiplier statistic', 'p-value',
        'f-value', 'f p-value']
test = sms.het_breuschpagan(res.resid, res.model.exog)
lzip(names, test)

[('Lagrange multiplier statistic', 8.034321223524866),
 ('p-value', 0.1543542336785945),
 ('f-value', 1.6072089552256548),
 ('f p-value', 0.15437651297930513)]

p-value = 0.15
Наш p-value значительно больше, чем 0.01 или 0.05, так как равен 0.15, следовательно, мы не отвергаем нулевую гипотезу о наличии гомоскедастичности в наших данных

## Задание 11

**Как вы считаете, будет ли в такой модели возникать проблема эндогенности? Почему?**

In [66]:
pip install linearmodels

  Attempting uninstall: typing-extensions
    Found existing installation: typing-extensions 3.10.0.2
    Uninstalling typing-extensions-3.10.0.2:
      Successfully uninstalled typing-extensions-3.10.0.2
  Attempting uninstall: Cython
    Found existing installation: Cython 0.29.24
    Uninstalling Cython-0.29.24:
      Successfully uninstalled Cython-0.29.24
Note: you may need to restart the kernel to use updated packages.


In [70]:
!pip install typing_extensions==4.4.0
import typing_extensions
from importlib import reload
reload(typing_extensions)
!pip install linearmodels

  Attempting uninstall: typing-extensions
    Found existing installation: typing-extensions 4.8.0
    Uninstalling typing-extensions-4.8.0:
      Successfully uninstalled typing-extensions-4.8.0


In [71]:
import linearmodels

In [72]:

instr = pd.get_dummies(df2, columns=["region"], drop_first=True)
a = instr.columns[9:48]

for i in range(37):
    instr[a[i]] = instr[a[i]].astype('int')
    
instr   

b = instr.iloc[:,9:48]
endog1 = df2['gender_ЖЕНСКИЙ']
dep = df2['ln_wage']
df2['sq_age'] = df2['sq_age'].astype("float")
df2['higher_educ_1.0'] = df2['higher_educ_1.0'].astype("float")
df2['married_1.0'] = df2['married_1.0'].astype("float")
exog = df2[['sq_age', 'higher_educ_1.0', 'married_1.0']]
exog = sm.add_constant(exog)
import linearmodels
from linearmodels import IV2SLS
ress = IV2SLS(dep,exog, endog1, b).fit(cov_type='unadjusted')
ress.summary
exog2 = df2[['gender_ЖЕНСКИЙ', 'sq_age', 'higher_educ_1.0', 'married_1.0']]
exog2 = sm.add_constant(exog2)
resultsOLS = IV2SLS(dep,exog2,None,None).fit(cov_type='unadjusted')
resultsOLS.summary
from collections import OrderedDict
from linearmodels.iv.results import compare
res1= OrderedDict()
res1['OLS'] = resultsOLS
res1['2SLS'] = ress
compare(res1)
ress.durbin()


C:\Users\Honor\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


Durbin test of exogeneity
H0: All endogenous variables are exogenous
Statistic: 57.0306
P-value: 0.0000
Distributed: chi2(1)
WaldTestStatistic, id: 0x246c6224bb0

Нулевая гипотеза данного теста состоит в том, что факторы модели являются экзогенными, однако, наш p-value = 0.00, что меньше, чем 0.01 или 0.05 (популярные уровни значимости), а значиит, мы отвергаем нулевую гипотезу, эндогенность наблюдается.